### PR TITLE
added e2e test for starvation

### PR DESCRIPTION
**Pull request by pdgetrf**
Added e2e for starvation. 

The fake starvation timeout is subject to tune. It should be set long enough to allow the first backfill job to run, and not too long to waste time.

Known issue: sometimes a deleted pod can be stuck in Terminating state, cause timeout and fail the test.

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -90,6 +90,7 @@ func (pc *Scheduler) Run(stopCh <-chan struct{}) {
 		panic(err)
 	}
 
+	glog.V(4).Infof("pc setting %+v", pc)
```
Comment (dhzhuo) [inline pkg/scheduler/scheduler.go:93]: Change to "scheduler configuration"?

File: test/e2e/job.go [hunk 2]
```diff
@@ -417,6 +417,106 @@ var _ = Describe("Job E2E Test", func() {
 		checkError(context, err)
 	})
 
+	FIt("Starvation prevention", func() {
+		context := initTestContext()
+		defer cleanupTestContext(context)
+		maxPods := clusterSize(context, oneCPU)
+
+		// create a job to prevent job "big" from running
+		replicaset := createReplicaSet(context, "rs1", maxPods-2, "nginx", oneCPU)
```
Comment (dhzhuo) [inline test/e2e/job.go]: Do you still plan to replace replicaset with job?

**Review by dhzhuo — APPROVED** [review]
Looks good. Just a couple of minor things.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 3]
```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
+				// allow over-allocate for starving job 
+				isJobStarving := job.Starving(ssn.StarvationThreshold)
+				toOverAllocateForStarv := false
+				if isJobStarving {
+					netResource := node.Allocatable.Clone()
+					for _, nodeTask := range node.Tasks {
+						if nodeTask.Job == job.UID &&
+						(nodeTask.Status == api.OverOccupied || nodeTask.Job == job.UID && nodeTask.Status == api.Allocated) {
+							netResource.Sub(nodeTask.InitResreq)
+							glog.Infof("removing %v of task %s from node %s", nodeTask.InitResreq, nodeTask.Name, node.Name)
```
Comment (dhzhuo) [inline pkg/scheduler/actions/allocate/allocate.go]: Can you give an explanation in log message on why you are removing tasks here?

File: pkg/scheduler/actions/allocate/allocate.go [hunk 4]
```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
 
-				if task.InitResreq.LessEqual(node.GetAccessibleResource()) {
-					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>",
-						task.Namespace, task.Name, node.Name)
+				// allow over-allocate for starving job 
+				isJobStarving := job.Starving(ssn.StarvationThreshold)
+				toOverAllocateForStarv := false
+				if isJobStarving {
+					netResource := node.Allocatable.Clone()
+					for _, nodeTask := range node.Tasks {
```
Comment (dhzhuo) [inline pkg/scheduler/actions/allocate/allocate.go]: This for loop will be executed for each task which is inefficient. We need to do a better job here. How about maintaining net resource inside NodeInfo?

File: pkg/scheduler/actions/allocate/allocate.go [hunk 5]
```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 				glog.V(3).Infof("Considering Task <%v/%v> on node <%v>. Task request: <%v>; Idle: <%v>; Used: <%v>; Releasing: <%v>; Backfilled: <%v>",
 					task.Namespace, task.Name, node.Name, task.Resreq, node.Idle, node.Used, node.Releasing, node.Backfilled)
 
-				if task.InitResreq.LessEqual(node.GetAccessibleResource()) {
-					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>",
-						task.Namespace, task.Name, node.Name)
+				// allow over-allocate for starving job 
+				isJobStarving := job.Starving(ssn.StarvationThreshold)
```
Comment (dhzhuo) [inline pkg/scheduler/actions/allocate/allocate.go]: Since we need to put starvation code in a separate PR, I would suggest push this commit to a different branch and leave it there. Let's first clean up backfill code based on review feedbacks, have it merged, and then work on the starvation. What do you think?

File: pkg/scheduler/actions/allocate/allocate.go [hunk 6]
```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
+						(nodeTask.Status == api.OverOccupied || nodeTask.Job == job.UID && nodeTask.Status == api.Allocated) {
+							netResource.Sub(nodeTask.InitResreq)
+							glog.Infof("removing %v of task %s from node %s", nodeTask.InitResreq, nodeTask.Name, node.Name)
+						}
+					}
+
+					// toOverAllocateForStarv == true means when allocating for starving job, over-allocate resources ignoring 
+					// resources being used on the node to prevent other non-startving job from taking the resources
+					toOverAllocateForStarv = (isJobStarving &&
+					!task.InitResreq.LessEqual(node.GetAccessibleResource()) &&
```
Comment (dhzhuo) [inline pkg/scheduler/actions/allocate/allocate.go]: I think the task can be allocated as long as its request is less than net resource. Why do you care if it is greater than accessible resource?

File: pkg/scheduler/actions/allocate/allocate.go [hunk 7]
```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
-				if task.InitResreq.LessEqual(node.GetAccessibleResource()) {
-					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>",
-						task.Namespace, task.Name, node.Name)
+				// allow over-allocate for starving job 
+				isJobStarving := job.Starving(ssn.StarvationThreshold)
+				toOverAllocateForStarv := false
+				if isJobStarving {
+					netResource := node.Allocatable.Clone()
+					for _, nodeTask := range node.Tasks {
+						if nodeTask.Job == job.UID &&
```
Comment (dhzhuo) [inline pkg/scheduler/actions/allocate/allocate.go]: I do not understand this if condition.  This if condition is pretty complicated. Can we simplify it?

File: pkg/scheduler/actions/allocate/allocate.go [hunk 8]
```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
+					task.InitResreq.LessEqual(netResource))
+				}
+
+				if task.InitResreq.LessEqual(node.GetAccessibleResource()) || toOverAllocateForStarv {
+					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>", task.Namespace, task.Name, node.Name)
+
+					usingBackfillTaskRes := !task.InitResreq.LessEqual(node.Idle) && !isJobStarving
 
-					if err := ssn.Allocate(task, node.Name, !task.InitResreq.LessEqual(node.Idle)); err != nil {
+					if err := ssn.Allocate(task, node.Name, usingBackfillTaskRes, toOverAllocateForStarv); err != nil {
```
Comment (dhzhuo) [inline pkg/scheduler/actions/allocate/allocate.go]: Try not passing bool into ssn.Allocate. It starts to get ugly. :)

File: pkg/scheduler/api/types.go [hunk 9]
```diff
@@ -29,6 +29,10 @@ const (
 	// and only if, N.IdleResource < T.RequestResource <= N.AllocatableResource.
 	AllocatedOverBackfill
 
+	// OverOccupied means the job is allocated even when there is no immedicately
+	// available resources due to starvation
+	OverOccupied
```
Comment (dhzhuo) [inline pkg/scheduler/api/types.go:34]: Do we need this state? How about we leave the job in `Allocated` state?

Comment (pdgetrf) [reply pkg/scheduler/api/types.go:34]: this state is to keep the task from being dispatched

Comment (dhzhuo) [reply pkg/scheduler/api/types.go:34]: how about merge `AllocatedOverBackfill` and `OverOcupied` into one `Waiting` state?

File: test/e2e/job.go (hunk 2 above)
Comment (pdgetrf) [reply test/e2e/job.go]: it's replaced

File: pkg/scheduler/actions/allocate/allocate.go (hunk 3 above)
Comment (pdgetrf) [reply pkg/scheduler/actions/allocate/allocate.go]: This is only removing for calculation purpose. this log is confusing to say the least. I'm removing this log.

File: pkg/scheduler/actions/allocate/allocate.go (hunk 6 above)
Comment (pdgetrf) [reply pkg/scheduler/actions/allocate/allocate.go]: it is to make sure toOverAllocateForStarv is set to true ONLY when the node does not have enough resource at the moment but can be allocated nonetheless because there's enough net resource

File: pkg/scheduler/api/types.go (hunk 9 above)
Comment (pdgetrf) [reply pkg/scheduler/api/types.go]: looking into it

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (pdgetrf) [reply pkg/scheduler/scheduler.go:93]: removed

File: pkg/scheduler/actions/allocate/allocate.go (hunk 5 above)
Comment (pdgetrf) [reply pkg/scheduler/actions/allocate/allocate.go]: as discussed yesterday. we'll do refactor, and then sync with the up stream and then separate PRs (I still have doubt about the need to separate but we'll see as we go)

File: pkg/scheduler/api/types.go (hunk 9 above)
Comment (pdgetrf) [reply pkg/scheduler/api/types.go:34]: doesn't look like they can be merged. AllocatedOverBackfill is also used for patching the backfill annotation. OverOcupied specifically means this task is allocated using without being able to be dispatched because resource is over-allocated for starvation prevention.

File: pkg/scheduler/actions/allocate/allocate.go (hunk 7 above)
Comment (pdgetrf) [reply pkg/scheduler/actions/allocate/allocate.go]: fixed.

File: pkg/scheduler/actions/allocate/allocate.go (hunk 8 above)
Comment (pdgetrf) [reply pkg/scheduler/actions/allocate/allocate.go]: being refactored

File: pkg/scheduler/api/types.go (hunk 9 above)
Comment (dhzhuo) [reply pkg/scheduler/api/types.go:34]: It looks to me that we use `task.IsBackfill` to decide whether or not to patch a pod. We do not use `AllocatedOverBackfill`.

File: pkg/scheduler/actions/backfill/backfill.go [hunk 10]
```diff
@@ -133,10 +134,35 @@ func backFill(ssn *framework.Session, job *api.JobInfo) {
 			}
 
 			if task.InitResreq.LessEqual(node.Idle) {
+
+				/* patch the annotation */
+				annotation := map[string]string{v1alpha1.BackfillAnnotationKey: "true"}
```
Comment (dhzhuo) [inline pkg/scheduler/actions/backfill/backfill.go]: I think we need to check the job status here. If job is not ready, we probably should not patch the pod.

We patch and bind pod only if the job is ready.

File: pkg/scheduler/actions/backfill/backfill.go [hunk 11]
```diff
@@ -133,10 +134,35 @@ func backFill(ssn *framework.Session, job *api.JobInfo) {
...
+				err := ssn.PatchAnnotation(task, annotation)
+				if err != nil {
+					glog.Errorf("Failed to patch backfill=true to task/pod <%s/%s>: %v, will not backfill on this node %s",
+						task.Name, task.Pod.Name, err, node.Name)
+					continue
```
Comment (dhzhuo) [inline pkg/scheduler/actions/backfill/backfill.go]: We have a tough question here. Suppose the job have 5 tasks. What should we do, if we successfully patch the first 3 tasks, but the patch fails on the 4th task?  

I think we need to make sure that if patch fails on any task, we will not proceed to bind any task. In this case, some tasks may have annotations, but no task will be bound. For tasks that have annotations but are not bound to any host, we can remove their annotation in the next scheduling round when we create TaskInfo from Pod.

File: pkg/scheduler/actions/backfill/backfill.go [hunk 12]
```diff
@@ -133,10 +134,35 @@ func backFill(ssn *framework.Session, job *api.JobInfo) {
...
+
+				glog.Infof("pod %s is marked as a backfill", task.Pod.Name)
 				task.IsBackfill = true
+
 				glog.V(3).Infof("Binding backfill task <%v/%v> to node <%v>", task.Namespace, task.Name, node.Name)
 				if err := ssn.Allocate(task, node.Name, false, false); err != nil {
 					glog.Errorf("Failed to bind backfill task %v on %v in Session %v: %s", task.UID, node.Name, ssn.UID, err)
+
+					// when task cannot be allocated, remove the backfill annotation if present
+					if task.IsBackfill == true {
```
Comment (dhzhuo) [inline pkg/scheduler/actions/backfill/backfill.go]: I feel that it is better to put this logic in `NewTaskInfo`. If Pod.hostname is NULL, we remove the annotation. What do you think?

**Review by dhzhuo — CHANGES_REQUESTED** [review]

File: pkg/scheduler/actions/backfill/backfill.go (hunk 11 above)
Comment (pdgetrf) [reply pkg/scheduler/actions/backfill/backfill.go]: that brings up another issue. what if dispatch fails on the 4th task?

Comment (dhzhuo) [reply pkg/scheduler/actions/backfill/backfill.go]: According to the current implementation, if dispatch fails on the 4th task, the scheduler will error out without dispatching the 5th task. 

Whether or not this is the correct behavior is debatable. But, at least, our patch logic does not make things worse. We may end up having a few tasks which have annotations but are not bounded. But this is okay because we are able to fix it in the next scheduling round.

File: pkg/scheduler/api/job_info.go [hunk 13]
```diff
@@ -79,7 +79,15 @@ func CheckBackfill(pod *v1.Pod) bool {
 				return false
 			}
 			glog.Infof("Restored backfill status for pod %s", pod.Name)
-			return hasBackfillAnnotation && pod.Status.Phase == v1.PodRunning
+
+			isPodScheduled := false
```
Comment (dhzhuo) [inline pkg/scheduler/api/job_info.go:83]: Why not use `len(pod.Spec.NodeName) == 0`?

File: pkg/scheduler/api/types.go (hunk 9 above)
Comment (dhzhuo) [inline pkg/scheduler/api/types.go:34]: Can you please update the comment as well? thansk
